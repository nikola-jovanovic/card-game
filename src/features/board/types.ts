import { TypeOf } from 'io-ts'

import { Card as CardC } from './entities/Card'

export type Card = TypeOf<typeof CardC>
export type Score = number

export enum Names {
  Me = 'Me',
  Milisav = 'Milisav',
  Mileva = 'Mileva',
  Djura = 'Djura',
}

export interface Player {
  name: Names
  score: Score
  cards: Card[]
  active: boolean
}

export type Pile = {
  [key in Names]?: Card
}