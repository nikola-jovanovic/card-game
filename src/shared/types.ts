export type Action<P, M = void> = {
  type: string
  payload: P
  meta: M
}

