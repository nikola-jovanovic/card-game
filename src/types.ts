import { TypeOf } from 'io-ts'
import { Card as CardC } from './entities/Card'

export type Action<P, M = void> = {
  type: string
  payload: P
  meta: M
}

export type Card = TypeOf<typeof CardC>
