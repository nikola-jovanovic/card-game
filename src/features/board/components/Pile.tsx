import React, { memo } from 'react'
import styled from 'styled-components'

import Centered from '../../../shared/components/Centered'
import { Pile as PileT } from '../types'
import Card from './Card'

export type Props = {
  pile: PileT
}

const Wrapper = styled.div`
  display: block;
  width: 100%;
  text-align: center;
`

const FlexContainer = styled.div`
  display: flex;
  justify-content: space-between;
`

const Pile = ({ pile }: Props): JSX.Element => {
  return (
    <Centered>
      <Wrapper>
        <div>
          {pile.Milisav && <Card {...pile.Milisav} />}
        </div>
        <FlexContainer>
          <div>
            {pile.Mileva && <Card {...pile.Mileva} />}
          </div>
          <div>
            {pile.Djura && <Card {...pile.Djura} />}
          </div>
        </FlexContainer>
        <div>
          {pile.Me && <Card {...pile.Me} />}
        </div>
      </Wrapper>
    </Centered>
  )
}

export default memo(Pile)
