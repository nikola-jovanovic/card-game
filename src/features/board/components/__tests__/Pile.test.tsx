import React from 'react'
import { render, screen } from '@testing-library/react'

import Pile, { Props } from '../Pile'
import { Names } from '../../types'

describe('Pile', () => {
  const props: Props = {
    pile: {
      [Names.Me]: { image: 'image', value: 12, code: 'JH' },
      [Names.Milisav]: { image: 'image', value: 13, code: 'QH' },
    }
  }
  const third = {
    [Names.Mileva]: { image: 'image', value: 14, code: 'KH' }
  }
  const fourth = {
    [Names.Djura]: { image: 'image', value: 1, code: 'AH' }
  }

  it('2 players', () => {
    const { container } = render(<Pile {...props} />)

    expect(screen.getAllByRole('img').length).toEqual(2)
    expect(screen.getByRole('img', { name: /jh/i })).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /qh/i })).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('3 players', () => {
    const newProps = {
      pile: {
        ...props.pile,
        ...third,
      }
    }
    const { container } = render(<Pile {...newProps} />)

    expect(screen.getAllByRole('img').length).toEqual(3)
    expect(screen.getByRole('img', { name: /jh/i })).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /qh/i })).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /kh/i })).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('4 players', () => {
    const newProps = {
      pile: {
        ...props.pile,
        ...third,
        ...fourth
      }
    }
    const { container } = render(<Pile {...newProps} />)

    expect(screen.getAllByRole('img').length).toEqual(4)
    expect(screen.getByRole('img', { name: /jh/i })).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /qh/i })).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /kh/i })).toBeInTheDocument()
    expect(screen.getByRole('img', { name: /ah/i })).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })
})