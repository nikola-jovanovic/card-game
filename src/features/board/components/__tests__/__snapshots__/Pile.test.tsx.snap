// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pile 2 players 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 10px;
}

.c2 {
  max-width: 80px;
}

.c1 {
  display: block;
  width: 100%;
  text-align: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div>
        <img
          aria-label="QH"
          class="c2"
          src="image"
        />
      </div>
      <div
        class="c3"
      >
        <div />
        <div />
      </div>
      <div>
        <img
          aria-label="JH"
          class="c2"
          src="image"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Pile 3 players 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 10px;
}

.c2 {
  max-width: 80px;
}

.c1 {
  display: block;
  width: 100%;
  text-align: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div>
        <img
          aria-label="QH"
          class="c2"
          src="image"
        />
      </div>
      <div
        class="c3"
      >
        <div>
          <img
            aria-label="KH"
            class="c2"
            src="image"
          />
        </div>
        <div />
      </div>
      <div>
        <img
          aria-label="JH"
          class="c2"
          src="image"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Pile 4 players 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 10px;
}

.c2 {
  max-width: 80px;
}

.c1 {
  display: block;
  width: 100%;
  text-align: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div>
        <img
          aria-label="QH"
          class="c2"
          src="image"
        />
      </div>
      <div
        class="c3"
      >
        <div>
          <img
            aria-label="KH"
            class="c2"
            src="image"
          />
        </div>
        <div>
          <img
            aria-label="AH"
            class="c2"
            src="image"
          />
        </div>
      </div>
      <div>
        <img
          aria-label="JH"
          class="c2"
          src="image"
        />
      </div>
    </div>
  </div>
</div>
`;
